docs/proposals/container-registry-credentials-per-usercluster.md

```diff
@@ -0,0 +1,207 @@
+  * [Seed Cluster](#seed-cluster)
+* [Alternative implementation](#alternative-implementation)
+  * [Set imagePullSecrets on user cluster ServiceAccounts](#set-imagepullsecrets-on-user-cluster-serviceaccounts)
+* [Alternatives considered](#alternatives-considered)
+  * [Set imagePullSecrets through an Addon](#set-imagepullsecrets-through-an-addon)
+  * [Reconciliation of all Pods running on user clusters](#reconciliation-of-all-pods-running-on-user-clusters)
+  * [Extension of Addon templates with imagePullSecrets](#extension-of-addon-templates-with-imagepullsecrets)
+  * [A look ahead](#a-look-ahead)
+* [Task and effort](#task-and-effort)
+
+## Goals
+
+Defining separate container registry credentials per user cluster, so cluster owners can make use of
+their docker hub credentials to avoid image pull limits or to access private images.
+
+## Non-Goals
+
+Setting local mirrors or private container registries for KKP components or Addons.
+
+## Motivation and Background
+
+Based on the user story [#6231][] users want to be able to use their Docker Hub paid accounts to
+mitigate image pull limits. Setting this up cluster wide is currently not possible.
+
+### Current state
+
+Currently, it's possible to set `spec.imagePullSecret` in the `KubermaticConfiguration`. The
+dockerconfig-json defined on this key is used in the Kubermatic Operator where the master- and the
+seed-controller create a `dockerconfigjson`-secret respectively. When creating the deployments for
+KKP master and seed components this secret gets referenced as `imagePullSecret` in all of their
+pod specs.
+
+This dockerconfig has no effect on anything running on the user clusters, so currently KKP does not
+offer a way to manage container registry credentials to be used by KKP internal components, KKP
+Addons nor user deployed applications by default.
+
+### Proposed feature
+
+`Cluster` resources get a new field `imagePullSecrets` that resembles `imagePullSecrets` on other
+resources like Deployments or Pods. The secrets referenced here get synchronized into the user
+cluster and a mutating admission webhook injects them into every Pod that is created on that user
+cluster.
+
+For this first iteration the secrets referenced on `Cluster` resources need to be created manually.
+
+There is also the possibility to define default credentials in the `KubermaticConfiguration` at
+`spec.userCluster.imagePullSecret`. If defined, this will be used for any user cluster that has no
+explicit `imagePullSecrets` set.
+
+## Implementation
+
+### KubermaticConfiguration
+
+A new configuration field to set default fallback registry credentials for any user cluster where no
+credentials are explicitly selected. For consistency this could be in the format of
+`spec.imagePullSecret`:
+
+```yaml
+apiVersion: kubermatic.k8c.io/v1
+kind: KubermaticConfiguration
+metadata:
+  name: <<mykubermatic>>
+  namespace: kubermatic
+spec:
+  userCluster:
+    imagePullSecret: |-
```

Comment:
We should be sure that we filter the domain for the image and only mutate the `imagePullSecret` in the pod spec when there is a match

Reply:
On the pods we don't modify any secret. We merge the ones already set in the pod spec with those defined on the `Cluster` CR (and the one referenced in the `KubermaticConfiguration`).
`imagePullSecrets` is a list of secret references and the kubelet will only use those where the hostname matches. It's also fine to have multiple secrets for the same hostname, the kubelet will try all of them until it succeeds to pull the image.